lightningd/log.c

```diff
@@ -183,41 +184,61 @@ static u32 delete_threshold(enum log_level level)
 	abort();
 }
 
+/* Delete a log entry: returns how many now deleted */
+static size_t delete_entry(struct log_book *log, struct log_entry *i)
+{
+	log->mem_used -= mem_used(i);
+	log->num_entries--;
+	if (i->nc && --i->nc->count == 0)
+		tal_free(i->nc);
+	free(i->log);
+	tal_free(i->io);
+
+	return 1 + i->skipped;
+}
+
 static size_t prune_log(struct log_book *log)
 {
-	struct log_entry *i, *next;
-	size_t skipped = 0, deleted = 0, count = 0, max;
+	size_t skipped = 0, deleted = 0, count = 0, dst = 0, max, tail;
 
 	/* Never delete the last 10% (and definitely not last one!). */
-	max = log->num_entries * 9 / 10 - 1;
+	tail = log->num_entries / 10 + 1;
+	max = log->num_entries - tail;
 
-	list_for_each_safe(&log->log, i, next, list) {
-		if (count++ == max) {
-			i->skipped += skipped;
-			skipped = 0;
-			break;
-		}
+	for (count = 0; count < max; count++) {
+		struct log_entry *i = &log->log[count];
 
 		if (pseudorand(1000) > delete_threshold(i->level)) {
 			i->skipped += skipped;
 			skipped = 0;
+			/* Move down if necesary. */
+			log->log[dst++] = *i;
```

Comment:
So this is effectively a merge-sort with itself, skipping deleted entries, that's very interesting :slightly_smiling_face:

Reply:
Well, it's *cheap* :)